app/views/base.py

```diff
@@ -0,0 +1,216 @@
+"""View-related code common to the whole app."""
+
+import functools
+import re
+
+import django.http
+import django.utils.decorators
+import django.views
+
+import config
+import const
+import resources
+import site_settings
+import utils
+
+
+class BaseView(django.views.View):
+    """Base view class shared across the app."""
+
+    # This should be overridden by subclasses.
+    ACTION_ID = None
+
+    _GET_PARAMETERS = {
+        'lang': utils.strip,
+    }
+
+    def setup(self, request, *args, **kwargs):
+        """Sets up the handler.
+
+        Views aren't passed any request-specific information when they're
+        initialized, so you can't really do much in __init__. However, having a
+        function that gets called before dispatch is useful for all kinds of
+        things (in particular, it's useful to have a top-down function, where
+        the parent class functions run before those of subclasses). setup() is
+        essentially a substitute for __init__().
+
+        Args:
+            request (HttpRequest): The request object.
+            *args: Unused.
+            **kwargs: Arbitrary keyword arguments. Should include repository ID
+                (under the key 'repo') if applicable.
+        """
+        # pylint: disable=attribute-defined-outside-init
+        # TODO(nworden): don't forget to call super.setup here once we upgrade
+        # to Django 2.2.
+        del request, args  # unused
+
+        # Set up the parameters and read in the base set of parameters.
+        self.params = utils.Struct()
+        self.read_params(get_params=BaseView._GET_PARAMETERS)
+
+        # Set up env variable with data needed by the whole app.
+        self.env = utils.Struct()
```

Comment:
I played around with this for a while, but it's trickier than I expected, and I can't come up with a way to do it that doesn't feel over-engineered and/or too inflexible. In particular, a nice thing about Structs is it's no problem for subclass views to add their own data (and we could define utility functions to add data to them too if we wanted), but that gets complicated when you try to make it a class.
The seemingly obvious solution is to have the Env class be subclassed alongside the View subclasses, and I tried implementing it that way (I pushed it on a separate branch so you can see the comparison [here](https://github.com/nworden/personfinder/compare/djangoapp...nworden:djangoapp-envclass)). It does work, but I feel like doing it this way would cause us to lose a lot of flexibility. For example, if a subclass needed to define an env variable based on one of its own parameters (like the base class does with the lang parameter), it couldn't do it, because its params haven't been read in yet.
Do you have a better idea of how to implement it as a class?

Reply:
> Do you know how many of views define its own env fields? IMHO env should provide general information about the environment, so it shouldn't be customized for every view. Having BaseViewEnv and AdminBaseViewEnv sounds good, but maybe there should be only these two?

I think when we get to end-user-facing views, the parent class for them will need some settings of its own too (e.g., to set whether it's lite mode).

> Can you elaborate that the problem is? Since looks like you are able to access param with no problem here

The base-level Env class can access params, but if the AdminEnv class needed to access admin-specific params, it couldn't, because the AdminView setup function wouldn't have called read_params yet. That didn't matter for the admin views, but it'll be an issue for stuff like the `ui` param for end-user views. Also, more generally, I'm a little bit nervous that we might end up with other things where it'd be easier to initialize the env at the same time we're initializing the view to avoid duplicated code. So far I've only needed a part of what main.py's doing.

If the Env classes didn't have logic, and the field-setting was still left to View.setup, I wouldn't be concerned about this though. We could set up Env classes/subclasses that only allowed specific pre-named properties to avoid typo risks and make it easier to see what env objects contain, but still have their fields set by the View itself. What do you think?

In any case, I would prefer to handle all this separately and stick with Struct envs for this PR.

Reply:
> > Do you know how many of views define its own env fields? IMHO env should provide general information about the environment, so it shouldn't be customized for every view. Having BaseViewEnv and AdminBaseViewEnv sounds good, but maybe there should be only these two?
> 
> I think when we get to end-user-facing views, the parent class for them will need some settings of its own too (e.g., to set whether it's lite mode).

I see, then there can be more than 2, but the number is still limited (for each base class like AppBase, StaticBase, but not for every view), right?

> 
> > Can you elaborate that the problem is? Since looks like you are able to access param with no problem here
> 
> The base-level Env class can access params, but if the AdminEnv class needed to access admin-specific params, it couldn't, because the AdminView setup function wouldn't have called read_params yet. 

Oh I see. This particular issue can be workarounded if you make either `param` or `__GET_PARAMETERS` a method (instead of a variable), and let subclasses override it. This may be a good thing on its own because you can avoid an issue that the value of `self.params` is different depending on when you access it.

> That didn't matter for the admin views, but it'll be an issue for stuff like the `ui` param for end-user views. Also, more generally, I'm a little bit nervous that we might end up with other things where it'd be easier to initialize the env at the same time we're initializing the view to avoid duplicated code. So far I've only needed a part of what main.py's doing.

I agree that it can potentially complicate the initialization problem than it currently is.

> 
> If the Env classes didn't have logic, and the field-setting was still left to View.setup, I wouldn't be concerned about this though. We could set up Env classes/subclasses that only allowed specific pre-named properties to avoid typo risks and make it easier to see what env objects contain, but still have their fields set by the View itself. What do you think?

That sounds like a good idea. One reason I didn't like `env` to be a struct is that it is hard to know what fields it has. Its class definition would serve as a useful documentation of the fields.

> 
> In any case, I would prefer to handle all this separately and stick with Struct envs for this PR.

SGTM.

Reply:
> I see, then there can be more than 2, but the number is still limited (for each base class like AppBase, StaticBase, but not for every view), right?

Right.

> One reason I didn't like env to be a struct...

Are there other reasons (especially any that my proposal wouldn't address)?

Reply:
Another reason is that, in our original (webapp2) implementation, there's lengthy code which initializes `env` directly written in main.py:
https://github.com/google/personfinder/blob/master/app/main.py#L261-L445
which I wasn't quite happy with.

It was partially already resolved in the new implementation because it's in a better place and is shorter (at least for now). So that would be OK. If it gets longer again, we may factor it out into a utility function somewhere.